# ESG Data ETL

## Objectives

- Load and inspect the raw Public Company ESG Ratings dataset.
- Assess data quality, including missing values, duplicates, data types and inconsistent values.
- Clean and prepare the dataset for exploratory data analysis and Power BI.

## Inputs
- `data/raw/data.csv`
## Outputs

- `data/processed/esg_cleaned.csv`
## Additional Comments

- The raw dataset will remain unchanged.


---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\CodeInstitute\\public_company_esg_analysis\\jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\CodeInstitute\\public_company_esg_analysis'

# Section 1

# Extract - Load Raw ESG Data

In [8]:
import pandas as pd

df = pd.read_csv("data/raw/data.csv")

df.head()


,ticker,name,currency,exchange,industry,logo,weburl,environment_grade,environment_level,social_grade,...,governance_grade,governance_level,environment_score,social_score,governance_score,total_score,last_processing_date,total_grade,total_level,cik
0,dis,Walt Disney Co,USD,"NEW YORK STOCK EXCHANGE, INC.",Media,https://static.finnhub.io/logo/ef50b4a2b263c84...,https://thewaltdisneycompany.com/,A,High,BB,...,BB,Medium,510,316,321,1147,19-04-2022,BBB,High,1744489
1,gm,General Motors Co,USD,"NEW YORK STOCK EXCHANGE, INC.",Automobiles,https://static.finnhub.io/logo/9253db78-80c9-1...,https://www.gm.com/,A,High,BB,...,B,Medium,510,303,255,1068,17-04-2022,BBB,High,1467858
2,gww,WW Grainger Inc,USD,"NEW YORK STOCK EXCHANGE, INC.",Trading Companies and Distributors,https://static.finnhub.io/logo/f153dcda-80eb-1...,https://www.grainger.com/,B,Medium,BB,...,B,Medium,255,385,240,880,19-04-2022,BB,Medium,277135
3,mhk,Mohawk Industries Inc,USD,"NEW YORK STOCK EXCHANGE, INC.",Consumer products,https://static.finnhub.io/logo/26868a62-80ec-1...,https://mohawkind.com/,A,High,B,...,BB,Medium,570,298,303,1171,18-04-2022,BBB,High,851968
4,lyv,Live Nation Entertainment Inc,USD,"NEW YORK STOCK EXCHANGE, INC.",Media,https://static.finnhub.io/logo/1cd144d2-80ec-1...,https://www.livenationentertainment.com/,BBB,High,BB,...,B,Medium,492,310,250,1052,18-04-2022,BBB,High,1335258


---

# Section 2

# Data Inspection

In [9]:
df.shape

(722, 21)

In [10]:
df.columns

Index(['ticker', 'name', 'currency', 'exchange', 'industry', 'logo', 'weburl',
       'environment_grade', 'environment_level', 'social_grade',
       'social_level', 'governance_grade', 'governance_level',
       'environment_score', 'social_score', 'governance_score', 'total_score',
       'last_processing_date', 'total_grade', 'total_level', 'cik'],
      dtype='str')

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 722 entries, 0 to 721
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   ticker                722 non-null    str  
 1   name                  722 non-null    str  
 2   currency              722 non-null    str  
 3   exchange              722 non-null    str  
 4   industry              709 non-null    str  
 5   logo                  702 non-null    str  
 6   weburl                706 non-null    str  
 7   environment_grade     722 non-null    str  
 8   environment_level     722 non-null    str  
 9   social_grade          722 non-null    str  
 10  social_level          722 non-null    str  
 11  governance_grade      722 non-null    str  
 12  governance_level      722 non-null    str  
 13  environment_score     722 non-null    int64
 14  social_score          722 non-null    int64
 15  governance_score      722 non-null    int64
 16  total_score        

In [12]:
df.isnull().sum()

ticker                   0
name                     0
currency                 0
exchange                 0
industry                13
logo                    20
weburl                  16
environment_grade        0
environment_level        0
social_grade             0
social_level             0
governance_grade         0
governance_level         0
environment_score        0
social_score             0
governance_score         0
total_score              0
last_processing_date     0
total_grade              0
total_level              0
cik                      0
dtype: int64

In [13]:
df.duplicated().sum()

np.int64(0)

---

# Section 3

## Data Cleaning and Transformation


### Remove Non-Analytical Columns

The `logo` and `weburl` columns contain company URLs and are not required for the ESG analysis. Therefore, they are removed from the analytical dataset. The original raw dataset remains unchanged.


In [14]:
df = df.drop(columns=["logo", "weburl"])

df.shape


(722, 19)

### Handle Missing Industry Values

The `industry` column contains 13 missing values. Since industry is important for ESG analysis, these records are retained and the missing values are labelled as `Unknown`.

In [15]:
df["industry"] = df["industry"].fillna("Unknown")

df["industry"].isnull().sum()


np.int64(0)

### Convert Date Column

The `last_processing_date` column is converted from text to datetime format to ensure the date is stored in an appropriate data type.


In [17]:
df["last_processing_date"] = pd.to_datetime(
    df["last_processing_date"],
    format="%d-%m-%Y",
    errors="coerce"
)

df["last_processing_date"].dtype


dtype('<M8[us]')

In [18]:
df["last_processing_date"].isnull().sum()

np.int64(0)

In [19]:
for column in [
    "environment_level",
    "social_level",
    "governance_level",
    "total_level"
]:
    print(column, df[column].unique())


environment_level <StringArray>
['High', 'Medium', 'Excellent']
Length: 3, dtype: str
social_level <StringArray>
['Medium', 'High', 'Low', 'Excellent']
Length: 4, dtype: str
governance_level <StringArray>
['Medium', 'High', 'Low']
Length: 3, dtype: str
total_level <StringArray>
['High', 'Medium']
Length: 2, dtype: str


In [20]:
for column in [
    "environment_grade",
    "social_grade",
    "governance_grade",
    "total_grade"
]:
    print(column, df[column].unique())


environment_grade <StringArray>
['A', 'B', 'BBB', 'BB', 'AA']
Length: 5, dtype: str
social_grade <StringArray>
['BB', 'B', 'A', 'BBB', 'CCC', 'AA']
Length: 6, dtype: str
governance_grade <StringArray>
['BB', 'B', 'BBB', 'C']
Length: 4, dtype: str
total_grade <StringArray>
['BBB', 'BB', 'A', 'B']
Length: 4, dtype: str


---

# Section 4

## Data Quality Validation

The cleaned dataset is validated to confirm that there are no remaining missing values or duplicate records and that the dataset is ready for analysis.


In [21]:
print("Dataset shape:", df.shape)
print("Total missing values:", df.isnull().sum().sum())
print("Duplicate rows:", df.duplicated().sum())


Dataset shape: (722, 19)
Total missing values: 0
Duplicate rows: 0


---

# Section 5

## Export Cleaned ESG Data

The cleaned dataset is exported as a new CSV file for further analysis and use in Power BI. The original raw dataset remains unchanged.


In [22]:
import os

os.makedirs("data/cleaned", exist_ok=True)

df.to_csv(
    "data/cleaned/esg_cleaned.csv",
    index=False
)

print("Cleaned dataset exported successfully.")


Cleaned dataset exported successfully.


## Final ETL Check

The exported cleaned dataset is reloaded to confirm that it was saved successfully and is ready for analysis.


In [23]:
df_cleaned = pd.read_csv("data/cleaned/esg_cleaned.csv")

print("Cleaned dataset shape:", df_cleaned.shape)
df_cleaned.head()


Cleaned dataset shape: (722, 19)


,ticker,name,currency,exchange,industry,environment_grade,environment_level,social_grade,social_level,governance_grade,governance_level,environment_score,social_score,governance_score,total_score,last_processing_date,total_grade,total_level,cik
0,dis,Walt Disney Co,USD,"NEW YORK STOCK EXCHANGE, INC.",Media,A,High,BB,Medium,BB,Medium,510,316,321,1147,2022-04-19,BBB,High,1744489
1,gm,General Motors Co,USD,"NEW YORK STOCK EXCHANGE, INC.",Automobiles,A,High,BB,Medium,B,Medium,510,303,255,1068,2022-04-17,BBB,High,1467858
2,gww,WW Grainger Inc,USD,"NEW YORK STOCK EXCHANGE, INC.",Trading Companies and Distributors,B,Medium,BB,Medium,B,Medium,255,385,240,880,2022-04-19,BB,Medium,277135
3,mhk,Mohawk Industries Inc,USD,"NEW YORK STOCK EXCHANGE, INC.",Consumer products,A,High,B,Medium,BB,Medium,570,298,303,1171,2022-04-18,BBB,High,851968
4,lyv,Live Nation Entertainment Inc,USD,"NEW YORK STOCK EXCHANGE, INC.",Media,BBB,High,BB,Medium,B,Medium,492,310,250,1052,2022-04-18,BBB,High,1335258


---

# Section 6

## ETL Summary

The raw ESG dataset was successfully extracted, inspected, cleaned, transformed, validated, and prepared for analysis.

- The original dataset contained 722 rows and 21 columns.
- Missing values and duplicate records were inspected.
- Unnecessary columns (`logo` and `weburl`) were removed.
- Missing `industry` values were labelled as `Unknown` to retain the records.
- The `last_processing_date` column was converted to datetime format.
- ESG grade and level categories were checked for consistency.
- Data quality was validated after cleaning and transformation.
- The cleaned dataset contains 722 rows and 19 columns.
- The processed dataset was exported as `esg_cleaned.csv` and is ready for EDA and Power BI analysis.
